File: docs/breez-sdk/snippets/swift/BreezSdkSnippets/Sources/ReserveLeaves.swift
import BigNumber
import BreezSdkSpark

func prepareSendPaymentReserveLeaves(sdk: BreezSdk) async throws {
    // ANCHOR: prepare-send-payment-reserve-leaves
    let paymentRequest = "<payment request>"
    let amountSats: BInt? = BInt(50_000)

    let prepareResponse = try await sdk.prepareSendPayment(
        request: PrepareSendPaymentRequest(
            paymentRequest: paymentRequest,
            amount: amountSats,
            tokenIdentifier: nil,
            conversionOptions: nil,
            feePolicy: nil,
            reserveLeaves: true
        ))

    // The reservation ID can be used to cancel the reservation if needed
    if let reservationId = prepareResponse.reservationId {
        print("Reservation ID: \(reservationId)")
    }

    // Send payment as usual using the prepare response
    // try await sdk.sendPayment(request: SendPaymentRequest(prepareResponse: prepareResponse))
    // ANCHOR_END: prepare-send-payment-reserve-leaves
}

func cancelPrepareSendPayment(sdk: BreezSdk) async throws {
    // ANCHOR: cancel-prepare-send-payment
    let reservationId = "<reservation id from prepare response>"

    try await sdk.cancelPrepareSendPayment(
        request: CancelPrepareSendPaymentRequest(reservationId: reservationId))
    // ANCHOR_END: cancel-prepare-send-payment
}

File: docs/breez-sdk/snippets/flutter/lib/reserve_leaves.dart
import 'package:breez_sdk_spark_flutter/breez_sdk_spark.dart';

Future<PrepareSendPaymentResponse> prepareSendPaymentReserveLeaves(
    BreezSdk sdk) async {
  // ANCHOR: prepare-send-payment-reserve-leaves
  String paymentRequest = "<payment request>";
  BigInt? amountSats = BigInt.from(50000);

  final prepareResponse = await sdk.prepareSendPayment(
      request: PrepareSendPaymentRequest(
          paymentRequest: paymentRequest,
          amount: amountSats,
          tokenIdentifier: null,
          conversionOptions: null,
          feePolicy: null,
          reserveLeaves: true));

  // The reservation ID can be used to cancel the reservation if needed
  final reservationId = prepareResponse.reservationId;
  if (reservationId != null) {
    print("Reservation ID: $reservationId");
  }

  // Send payment as usual using the prepare response
  // sdk.sendPayment(request: SendPaymentRequest(prepareResponse: prepareResponse, options: null, idempotencyKey: null));
  // ANCHOR_END: prepare-send-payment-reserve-leaves
  return prepareResponse;
}

Future<void> cancelPrepareSendPayment(BreezSdk sdk) async {
  // ANCHOR: cancel-prepare-send-payment
  String reservationId = "<reservation id from prepare response>";

  await sdk.cancelPrepareSendPayment(
      request: CancelPrepareSendPaymentRequest(
          reservationId: reservationId));
  // ANCHOR_END: cancel-prepare-send-payment
}
